api/edge/v1alpha1/types.go

```diff
@@ -205,8 +205,9 @@ type PlacementDecisionSpec struct {
 }
 
 // DownsyncObjectReferences explicitly defines the objects to be down-synced.
-// The ClusterScope list defines the cluster-scope objects, while NamespacedObjects packs individual objects
-// identifiable by namespace & name.
+// The ClusterScope list defines the cluster-scope objects, NamespacedObjects packs individual objects
+// identifiable by namespace & name, and generation represents the internal spec generation of the listed
```

Comment:
(1) How does "internal space" differ from "spec"? The standard terminology is just "spec" --- or "desired state" if you mean something more than the spec section.

(2) Do you really mean that no changes to anything besides the spec section matter here? No changes to labels or annotations? No changes to ConfigMap or Secret objects?

Reply:
@MikeSpreitzer thanks for the comments. good questions. 
I will try to better explain. you're right that this comment doesn't explain the issue very good.
I need to update the comment (we do care about labels/annotations, not only spec).

the point we're tackling is how the update of an object is handled. 
Suppose you create a deployment in the WDS and a placement that selects this deployment. 
The placement controller writes the placement decision, the transport controller sees there is a new (or updated) placement decision, it wraps it up and deliver it to the mailbox namespace. 
At this point, the deployment may be updated - e.g., changing the number of replicas from 1 to 3. 
How the transport can informed of the change in spec of the deployment (or annotations or labels)? 

One valid option is to add the above field. 
I should point out that without this field, the PlacementDecision spec is not changed (cause list of objects and list of destinations remain the same), therefore placementdecision generation is not increased. 
with this field it triggers the increment of PlacementDecision.generation.
then we work with the desired state vs current state approach - 
PlacementDecision generation was increased (for example) to 4, while the placementdecision generation that was used to create the wrapped object is 3, so we need to update.
if both equal we don't need to update.
Think about it the following way - if in every object we would add the generation of it, would it be valid?
cause then we specify the concrete generation of the object to distribute which is the desired-state.

There is another valid approach which is to force the placement controller to update the placementdecision object even if nothing in the lists has changed, which will trigger the incrementation of the placementdecision.generation and then we do the same in the part of desired-state vs current-state.
wdyt?

Reply:
Does a no-change update increment `ObjectMeta.Generation`?

There is another valid approach, which I tried to outline above. In this approach the placement controller is only responsible for maintaining correct lists in the PlacementDecision, and the transport controller is responsible for maintaining the correct relationships between workload objects and wrapped-up contents in the mailboxes.

Reply:
If I understand correctly, in the approach suggested in this PR: (a) the placement controller has to update the ObjectsInternalSpecGeneration field of every PlacementDecision every time the placement controller starts up, and (b) the transport controller has to construct a fresh wrapped-up contents for every PlacementDecision object every time the transport controller starts up.

Reply:
```does a no-change update increment ObjectMeta.Generation? I would expect not.```
I'm pretty sure I tested it and there is a field to force the update, including the increment of generation.

option (3) is indeed another valid option, although it sounds to me like a more complex one.
I think options (1) or (2) are very simple and easy for understanding and maintaining, while option (3) might be trickier. 

regarding the startup - yes, we should trigger the update sequence. it might happen if any of the objects has changed or using this field.

Reply:
another comment on option (3) - 
in our design there is a clear separation of responsibilities between the placement controller and the transport controller.
placement controller is responsible for watching placement and all objects that are included in any of the placements. the placement controller should maintain the list of objects to distribute and list of destinations. upon any change in the objects or destinations, this controller should handle the change.

the transport controller is responsible for watching placement decision, fetch the objects as listed in the decision and distribute it to the destinations as listed in the decision object.

option (3) you suggested violates this separation of concerns. it's the placement controller responsibility to handle any change in placement or any of the objects the are selected by a placement.

Reply:
@MikeSpreitzer I want to correct myself regarding option (2). 
Maroon and me tested it today and seems like generation is not increasing. I'm pretty sure I was testing something with force and that increased generation, but maybe I had a bug or something that modified the spec and that was the reason for generation change.
so - forget about option (2), sounds like this is not an option :)
we're left with the other options.

Reply:
Since we are talking about options there are yet other two (let's call them option 4 and 5) which might be potentially more efficient. In the current implementation (kslight) as you know there is no creation or update of a PlacementDecision object (which 1) creates the issues described above to manage objects updates and 2) requires writes to the WDS API server for each object update). 

Option 4 - just implement the go interface defined for transport based on the work API and for every pluggable transport implement the code in a separate go project / package (as I think you are doing now) and import that code/package in the KubeStellar controller manager. Assuming we have other transport plugins, the selection of one or the other would be based on some flag on the KS controller manager. We may think this approach may make more difficult to add other plugins, but after all, how many do we expect. Most likely we will have the OCM-based one and the KS syncer-based one.

Option 5 - Expose the go transport interface as a gRPC interface. The transport would run in a separate container in the same pod as the placement controller or in a separate pod and exposed as a service. Whenever the placement controller sees an update on the tracked objects, it runs all the placement matching logic for objects and clusters and invokes the gRPC transport service which takes care of the transport part.

Reply:
@pdettori correct me if I'm wrong, but option (5) is kind of equal to option (1) in terms of triggering transport reconciliation from the placement controller. upon an internal change in one of the objects that are selected by a placement, option (5) uses gRPC to trigger reconciliation while option (1) increments a generation field in PlacementDecision object to trigger the reconciliation.

At the end, one of the main aspects I'm looking at is simplicity, and triggering reconciliation using a field in a CR sounds like to me like the most straight forward and simple approach.

Reply:
@nirrozenbaum: in terms of simplicity option 4 is the simplest of all (and probably most performant). Option 5 was adding the gPRC interface on the go interface you already defined for "work"  to avoid the tight coupling of the code. The gPRC interface is not about "triggering" the reconciliation but to do a gRPC call to the transport service that has all the objects reference and list of clusters as payload (basically the content of what you have in the PlacementDecision). The reconciliation happens on the client side (the placement controller) - if the gRPC call does not succeed it will re-enque and retry until succeeds.